vm/vm.go

```diff
@@ -702,28 +704,52 @@ func (vm *VM) Submit(
 	return errs
 }
 
-// populateValidityWindow populates the VM's time validity window on startup,
-// ensuring it contains recent transactions even if state sync is skipped (e.g., due to restart).
-// This is necessary because a node might restart with only a few blocks behind (or slightly ahead)
-// of the network, and thus opt not to trigger state sync. Without backfilling, the node's validity window
-// may be incomplete, causing the node to accept a duplicate transaction that the network already processed.
-
-// When Initialize is called, vm.consensusIndex is nil—it is set later via SetConsensusIndex.
-// Therefore, we must use the chainStore (which reads blocks from disk) to backfill the validity window.
-// This prepopulation ensures the validity window is complete, even if state sync is skipped.
-func (vm *VM) populateValidityWindow(ctx context.Context) error {
-	lastAcceptedBlkHeight, err := vm.chainStore.GetLastAcceptedHeight(ctx)
+// initValidityWindow populates the VM's time validity window on startup
+//
+// During initialization, there are two scenarios:
+//  1. First startup or state sync needed: Create an empty validity window that will be
+//     populated during bootstrapping/state sync.
+//  2. Normal operation or restart: Initialize with populated validity window using the
+//     last known block as the starting point.
+//
+// The populated validity window will use the last accepted block as its anchor point,
+// then scan the ancestry to build a complete validity window. This handles cases where
+// a node restarts with an incomplete validity window that could allow duplicate transactions.
+func (vm *VM) initValidityWindow(ctx context.Context) error {
+	getWindowSize := func(timestamp int64) int64 {
+		return vm.ruleFactory.GetRules(timestamp).GetValidityWindow()
+	}
+
+	// If we don't have a last accepted block or we're not in normal operation,
+	// initialize with empty validity window to be populated later
+	lastAcceptedHeight, err := vm.chainStore.GetLastAcceptedHeight(ctx)
+	normalOp := vm.normalOp.Load()
```

Comment:
If we move this into `startNormalOp` then we can guarantee that at least the genesis or some last accepted block has been accepted, so that we do not need to handle the error this way.